src/google/protobuf/descriptor.cc

```diff
@@ -5487,6 +5497,116 @@ void DescriptorBuilder::BuildMessage(const DescriptorProto& proto,
   }
 }
 
+void DescriptorBuilder::CheckFieldJsonNameUniqueness(
+    const DescriptorProto& proto, const Descriptor* result) {
+  // Two passes: one looking only at default JSON names, and one that considers
+  // custom JSON names. We can skip the first pass for proto2. (There may be some
+  // value in it, since FieldMask only uses default JSON names; but we skip it
+  // because it can be confusing, especially for fields that have custom names.)
+  FileDescriptor::Syntax syntax = result->file()->syntax();
+  std::string message_name = result->full_name();
+  if (syntax != FileDescriptor::SYNTAX_PROTO2) {
+    CheckFieldJsonNameUniqueness(message_name, proto, syntax, false);
+  }
+  CheckFieldJsonNameUniqueness(message_name, proto, syntax, true);
+}
+
+namespace {
+// Helpers for function below
+
+struct JsonNameDetails {
+  const FieldDescriptorProto* field;
+  std::string orig_name;
+  bool is_custom;
+};
+
+JsonNameDetails GetJsonNameDetails(const FieldDescriptorProto* field, bool use_custom) {
+  std::string default_json_name = ToJsonName(field->name());
+  if (use_custom && field->has_json_name() &&
+      field->json_name() != default_json_name) {
+    return {field, field->json_name(), true};
+  }
+  return {field, default_json_name, false};
+}
+
+bool JsonNameLooksLikeExtension(std::string name) {
+  return !name.empty() && name.front() == '[' && name.back() == ']';
+}
+
+
+} // namespace
+
+void DescriptorBuilder::CheckFieldJsonNameUniqueness(
+    const std::string& message_name, const DescriptorProto& message,
+    FileDescriptor::Syntax syntax, bool use_custom_names) {
+
+  absl::flat_hash_map<std::string, JsonNameDetails> name_to_field;
+  for (const FieldDescriptorProto& field : message.field()) {
+    JsonNameDetails details = GetJsonNameDetails(&field, use_custom_names);
+    if (details.is_custom && JsonNameLooksLikeExtension(details.orig_name)) {
+      std::string error_message =
+          absl::StrFormat("The custom JSON name of field \"%s\" (\"%s\") is invalid: "
+                          "JSON names may not start with '[' and end with ']'.",
+                          field.name(), details.orig_name);
+      AddError(message_name, field, DescriptorPool::ErrorCollector::NAME,
+               error_message);
+      continue;
+    }
+
+    std::string lowercase_name = absl::AsciiStrToLower(details.orig_name);
```

Comment:
Why do we do a case-insensitive search here?  AFAIK json is case-sensitive, so this is creating clashes where there don't necessarily need to be any right?

Reply:
I asked the same question when I opened this issue over two years ago: https://github.com/protocolbuffers/protobuf/issues/7192

I did not get a clear answer, so I didn't try to change that with this PR. The check for enum value names is also weirdly case-insensitive. 🤷 

I had assumed it was just to catch potentially confusing JSON names, that vary only in case. But, TBH (and I think I said this in that issue), that seems like a linter/style issue, not a compile issue. I'd be more than happy to make these checks more intuitive. They definitely feel overly broad as is today.

Reply:
I'll follow up with @acozzette about this, but I think if we remove this case-insensitivity it'll fix most (if not all) of the remaining issues

Reply:
Update: discussed this with @haberman and a case-sensitive comparison should be ok here.  As far as we can tell, this was made case-insensitive purely as a conservative measure.  I'm also investigating the fallout from taking this PR a step further and unconditionally erroring in both proto2 and proto3.  Will keep you updated